### PR TITLE
Add `String.reverse` method

### LINKED ISSUE #3503: Add `String.invert()`/`reverse()`
### Describe the project you are working on

Plugins

### Describe the problem or limitation you are having in your project

I need to invert a string but I can't to do this simply and efficient.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Adding `String.invert()`/`String.reverse()` would solve the problem

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The method would return new inverted string
`"hello".invert()` would give `"olleh"`

### If this enhancement will not be used often, can it be worked around with a few lines of script?

There are some workrounds, for example
```gdscript
static func invert_string(s:String)->String:
    var chars_pool = PoolStringArray()
    var length = s.length()
    chars_pool.resize(length)
    for i in length:
        chars_pool[i] = s[i]
    chars_pool.invert()
    return chars_pool.join("")
```
This doesn't look efficient tho

### Is there a reason why this should be core and not an add-on in the asset library?

Inverting strings is quite frequent task when working with text. For example in Python you can use `[::-1]` to do this.

### PR DESCRIPTION
**Pull request by Chaosus**
Adds a simple method to return the copy of the string in reverse order. May be useful in some cases. Closes https://github.com/godotengine/godot-proposals/issues/3503

### CODE REVIEW

**Review by dalexeev — APPROVED** [review]

**Comment by RedworkDE** [comment]
1. This implementation breaks some characters, eg `"🇺🇦".reverse() == "🇦🇺"`
2. Generally you would want to keep unicode grapheme clusters intact and just reverse their order, but in some locales a grapheme cluster is basically an entire word, so there it's probably not what you want either.
3. C# doesn't have this built-in because of exactly such problems, also its naive implementation would give different results.

TL/DR: this is a hard problem and I don't think this implementation does it justice.

**Comment by AThousandShips** [comment]
I think the quirks of reversing should be documented as per the above mentioned

**Comment by dalexeev** [comment]
@RedworkDE This is a good point, but note that some graphemes are composed of multiple code points:

```gdscript
print("🇺🇦".length()) # 2
```

I think we can just document that the `reverse()` method considers code points under individual characters, not graphemes (sorry if I misused the terms).

**Comment by AThousandShips** [comment]
Something like:
Note: This treats individual Unicode codepoints as individual characters, and may cause unexpected behaviour when reversing sequences including emoji, flags, etc.

**Comment by Zireael07** [comment]
I don't know a reverse() implementation that does NOT break flags/emoji and the like. Should be documented and that's all.

**Comment by bruvzg** [comment]
> This implementation breaks some characters, eg "🇺🇦".reverse() == "🇦🇺"

None of the other `String` functions will handle composite emojis as a single character, so this is not an issue. Besides this, the result reversing string won't make any sense in many cases (any context dependent character combinations, combining diacritics etc.).

**Comment by YuriSizov** [comment]
I agree that we shouldn't worry ourselves too much with how it handles complex unicode sequences and various composite languages. The resulting string won't be a valid word or sentence in practically any language. The application for this method lies somewhere else. Though the proposal doesn't elaborate why this is needed, only that it's needed.

If we seriously want to discuss if this is worth it or not, the proposal must provide some details for a use case of reversing a string, and doing so especially efficiently too to be considered a built-in (cc @me2beats). But we may also just YOLO this, as the naive implementation is pretty straightforward.

**Comment by me2beats** [comment]
I used string reverse mainly for searching replacing and sorting
with and without Regex

**Review by akien-mga — APPROVED** [review]
The implementation is straightforward, and it seems like a useful addition for the String class.

It's always hard to know what's the threshold for accepting new helper methods in core, but this seems OK to me, and nobody voiced any specific concerns in the past few weeks.

**Comment by akien-mga** [comment]
Thanks!
